### ISSUE TITLE
Doesn't recover from server down

### ISSUE THREAD

**Issue by scargill** [tzapu/node-red-contrib-blynk#16]
It SEEMS that the only way this recovers from the Blynk server being temporarily down - is to restart node-red... any chance of working on that?

**Comment by Toshibass** [comment]
+1 for that, I have same same critical issue, I am currently using Blynk cloud so when away from home if the blynk server "goes down" or " I lose connection or whatever"  Its not that easy to restart node-red.

**Comment by tzapu** [comment]
hi guys, any node red log files would be very helpful.

**Comment by scargill** [comment]
As soon as I am at a desk.. certainly

**Comment by tzapu** [comment]
have a good trip ;)

btw guys, i m experiencing the following issue randomly, resulting in a node red crash, maybe it is related. i get error

```
Assertion failed: (0), function uv_err_name, file ../deps/uv/src/uv-common.c, line 75.
```

have you seen anything similar? does your node red keep running (or being auto restarted) when the connection goes down, or is it dead as well?

**Comment by Toshibass** [comment]
I don't see any errors in the node-red log for issue #16 just red dot "not connected" on the node, the ONLY way to get the blynk nodes reconnected is to reboot node-red, then I get a green dot "connected" on the node.

Regarding random issue, crashing node-red, I had only 2 a month apart as follows:

26 Oct 16:23:18 - [red] Uncaught Exception:
26 Oct 16:23:18 - Error: write EPIPE
    at exports._errnoException (util.js:746:11)
    at WriteWrap.afterWrite (net.js:775:14)

22 Nov 12:41:02 - [info] Started modified flows  <--- 12.41pm (midday) was the last time I started a modified flow
23 Nov 01:39:50 - [red] Uncaught Exception:     <--- Exception happened 1.39am (not like 48 minutes after starting flows)
23 Nov 01:39:50 - Error: write EIO
    at exports._errnoException (util.js:746:11)
    at WriteWrap.afterWrite (net.js:775:14)

One final point I don't use pm2 to auto start/stop node-red currently, I use a .sh file activated in rc.local when the raspi boots... just a wondering if using pm2 would be advantageous.

Toshi

**Comment by scargill** [comment]
Well all I can say about pm2 is regardless of the odd crash in node red it 
ALWAYS restarts

**Comment by tzapu** [comment]
guys, this is the behaviour of the blynk js library mentained by the guys at blynk.
i will try to come up with a solution and issue a pull request to them.

**Comment by tzapu** [comment]
guys, created a pull request, as soon as it is merged this should be fixed after doing an update of the blynk-library package
https://github.com/vshymanskyy/blynk-library-js/pull/11

**Comment by Toshibass** [comment]
That's great, I had another crash this morning seams as soon as I tried to log on to node-red according to the times 

23 Nov 18:54:38 - [info] Started modified flows
24 Nov 09:27:26 - [red] Uncaught Exception:
24 Nov 09:27:26 - Error: write EIO
    at exports._errnoException (util.js:746:11)
    at WriteWrap.afterWrite (net.js:775:14)

I think I will try pm2 once I get time today.

**Comment by Toshibass** [comment]
Just saw the pull request, just to be clear @scargill is using local server and android I am using blynk cloud and Iphone so the issue appears to be a global one.

**Comment by tzapu** [comment]
hi, the pull request has been merged, but they also need to release a new version of the lib before it can be updated to
the fix will be a gobal one as well, it will even work for espruino or whatever else they support in their js  lib

**Comment by tzapu** [comment]
the pull request was accepted and the new version of the blynk lib released. if you update both that and node red contrib blynk you should have auto reconnect on server going away

**Comment by Toshibass** [comment]
Hi tzapu .. Ok I Just did a update.

(one point about that, I didn't install the blynk lib separately, I use the one that comes with node-red-contrib-blynk so I just ..sudo npm uninstall node-red-contrib-blynk then install it again) so currently 
"blynk-library",  "version": "0.0.24", &  "node-red-contrib-blynk",  "version": "0.0.17"

Re Node-red .. there was a maintenance release yesterday which solves the "EventEmitter memory leak detected. 11 listeners...." issue.

Obviously time will tell if the updated blynk lib will resolve this #16 issue, I will report back after a period of use, 

Thanks for your continued support.

Toshi

**Comment by scargill** [comment]
NOT solved.

Tonight – “your generic board is not in the network” – I went to Node-Red and the Blynk node had failed.  The App was getting history so server ok but it must’ve gone off at sometime during the evening. A reset of Node-Red (latest release) brought it back up – but the node is NOT recovering from server down.

Pete.

**Comment by tzapu** [comment]
@Toshibass "blynk-library", "version": "0.0.24", & "node-red-contrib-blynk", "version": "0.0.17" is right.
@scargill i need a way to replicate your server going down then... 
if you quit the blynk server app and restart it, does it reconnect alright ? 
maybe when it dies, it dies in a different kind of way

**Comment by scargill** [comment]
You are right. I’ve just done tests. Manually stopping and starting the Blynk server does NOT cause a problem. The whole thing recovers in a matter of several seconds. So it must be something else. I will wait for the next problem and send logs.

**Comment by scargill** [comment]
I found and fixed an issue with Node-red  - and so now my Node-Red/Pi installation runs without issue… however…

This morning  I noted that the Android App said my installation was offline – sure emough I looked at Node Red – everything working EXCEPT the Blynk nodes – all with red circles.   PM2 RESTART NODE-RED – and they were up and running…  so clearly there is still an issue where the node does NOT come back up after a problem even though the local BLYNK server is running..

Don’t know if this helps – BLYNK LOG

00:35:15.232 INFO  - pete@scargill.org app joined.

02:37:30.395 INFO  - pete@scargill.org app joined.

03:18:18.766 INFO  - pete@scargill.org app joined.

04:29:08.117 INFO  - pete@scargill.org app joined.

06:06:09.310 INFO  - pete@scargill.org app joined.

06:06:11.069 ERROR - User has exceeded message quota limit.

07:50:37.612 INFO  - pete@scargill.org app joined.

08:06:18.690 INFO  - pete@scargill.org app joined.

08:09:50.750 INFO  - pete@scargill.org app joined.

08:35:33.285 INFO  - pete@scargill.org app joined.

09:49:26.579 INFO  - pete@scargill.org app joined.

10:05:09.537 INFO  - pete@scargill.org app joined.

10:26:51.624 INFO  - pete@scargill.org app joined.

11:45:51.682 INFO  - pete@scargill.org app joined.

11:45:53.073 ERROR - User not logged. /81.174.144.64:44953. Closing.

11:45:55.632 INFO  - pete@scargill.org app joined.

11:48:06.295 INFO  - pete@scargill.org hardware joined.

From: Peter Scargill [mailto:pete@scargill.org] 
Sent: 02 December 2015 12:28
To: 'tzapu/node-red-contrib-blynk' reply@reply.github.com; 'tzapu/node-red-contrib-blynk' node-red-contrib-blynk@noreply.github.com
Subject: RE: [node-red-contrib-blynk] Doesn't recover from server down (#16)

You are right. I’ve just done tests. Manually stopping and starting the Blynk server does NOT cause a problem. The whole thing recovers in a matter of several seconds. So it must be something else. I will wait for the next problem and send logs.

**Comment by tzapu** [comment]
06:06:11.069 ERROR - User has exceeded message quota limit.
hi, this is a bit worrying. since it is your manual installation, could you increase the quota limit to something silly and see what happens?
cheers

**Comment by scargill** [comment]
A little more info would help – I’m not sure where I’m supposed to do that??

Pete.

**Comment by tzapu** [comment]
let me check.
i have seen something similar with blynk.cc and only when connecting the app to the server. it would cause a flood of requests on my esps which in turn kill the connection so they would have to reconnect

**Comment by tzapu** [comment]
from what they are saying here 
https://github.com/blynkkk/blynk-server under advanced setup, you have a property that you need to set 

```
user.message.quota.limit=100
```

as per example config file
https://github.com/blynkkk/blynk-server/blob/master/server/tcp-server/src/main/resources/server.properties

since this is a local instalation for you, you should not run into quota limitations at all
maybe it is linked

**Comment by scargill** [comment]
Right – I’ve added the file and taken that up to 250 max. Let’s see what happens and thanks for that. I’ll report back in a few days.

**Comment by tzapu** [comment]
really curiosu to see what will happen. if it were to default to 10 connections in the absence of that file, it would exaplain disconnecting node red
(although it doesn t yet explain why it wouldn t reconnect)

**Comment by scargill** [comment]
Well it is now set in 2 completely isolated installations – so with luck I’ll know in a matter of days if it has made a difference and will report back.

Pete

**Comment by scargill** [comment]
Not entirely sure what happened here as I’ve not changed anything and I’m not using secure…    This morning I noted that my my Node-Red installation in Spain (max increased to 250 as per earlier conversation)  showed Blynk as dead…. 

Here is the log..

07:25:35.185 INFO  - pete@scargill.org hardware joined.

07:25:35.190 INFO  - pete@scargill.org hardware joined.

07:26:56.900 INFO  - Using data dir '/var/blynk'

07:27:04.904 ERROR - Error parsing file '/var/blynk/server.properties'.

07:27:04.904 ERROR - Error parsing file '/var/blynk/admin-0.11.3.jar'.

07:27:04.904 ERROR - Error parsing file '/var/blynk/mail.properties'.

07:27:04.904 ERROR - Error parsing file '/var/blynk/server-0.11.3.jar'.

07:27:04.907 ERROR - Error parsing file '/var/blynk/server-0.11.3.jar.pid'.

07:27:07.268 INFO  - Initializing mail transport. Username : techscargill@gmail.com. SMTP host : smtp.gmail.com:587

07:27:13.256 INFO  - Plain tcp/ip hardware server port 8442.

07:27:13.264 WARN  - ATTENTION. Certificate /var/blynk and key /var/blynk paths not valid. Using embedded certs. This is not secure. Please replace it with your own certs.

07:27:19.894 INFO  - SSL hardware port 8441.

07:27:19.924 INFO  - Enabling SSL for application.

07:27:19.925 WARN  - ATTENTION. Server certificate paths cert : '/var/blynk', key : '/var/blynk' - not valid. Using embedded server certs and one way ssl. This is not secure. Please replace it with your own certs.

07:27:21.686 INFO  - Application server port 8443.

07:27:21.703 INFO  - Administration server port 8777.

07:27:43.689 INFO  - pete@scargill.org hardware joined.

A quick restart of Node-Red (pm2 restart node-red) – and all is well !!!

Any ideas?

**Comment by tzapu** [comment]
hi

this `07:27:04.904 ERROR - Error parsing file '/var/blynk/server.properties'.` would suggest it s not using the file. I don t know what all those errors mean really.

I am more interested anyway on the _node-red side of the log, if possible, when it disconnected_. I must setup a local server for myself i guess as i m not using one and let it running for a while.

**Comment by scargill** [comment]
Another one for you..

This morning the Blynk nodes were dead again – a quick restart of Node-Red again brought them back to life. I checked – the Pi has not reset for over 2 days (and that was deliberate). For a second I wondered if it was down to Node-Red starting before the Blynk server but that theory is now dead.

Nothing in any of the Blynk logs to suggest any problems.

Pete.

**Comment by tzapu** [comment]
hi, i d like to see the node-red log at the moment of disconnection
cheers

**Comment by scargill** [comment]
I have attached the node-red log file…

**Comment by tzapu** [comment]
hi @scargill , i may be slow (so excuse me in advance), but all I am seeing above are blynk server logs, not node red logs...

**Comment by scargill** [comment]
What I sent you last night were my node-red logs as created by PM2.

**Comment by tzapu** [comment]
It is not visible here:
https://github.com/tzapu/node-red-contrib-blynk/issues/16 
or on email notifications. Maybe it got ignored by github ?

**Comment by tzapu** [comment]
last i can see is

> 07:25:35.185 INFO  - pete@scargill.org hardware joined.
> 07:25:35.190 INFO  - pete@scargill.org hardware joined.
> 07:26:56.900 INFO  - Using data dir '/var/blynk'
> 07:27:04.904 ERROR - Error parsing file '/var/blynk/server.properties'.
> ...

which looks to me like blynk

**Comment by scargill** [comment]
Thanks I have logged this with BLYNK.    However I should point out that Blynk continues to run because every time the node goes down, restarting Node-Red restores operation.

I’ll let you know what Blynk say about below. Thanks.

**Comment by tzapu** [comment]
sorry, what i meant to say is that I have not received or seen any of your node-red logs, only blynk logs.

i have also setup a local blynk server for myself passing a lot of data to try and test this (see here https://github.com/tzapu/node-red-contrib-blynk/issues/20 )
been running for 24h now, with a write every 0.5 seconds, and a lot of updates at once when the app is open on the phone

**Comment by scargill** [comment]
Sorry…

Ok – erm 24 hours – you’ll need a couple of days – mine does not give up every 5 minutes :) maybe once a day – sometimes less, sometimes more. Ok,  I included my node-red log – is it possible the INCLUDE didn’t work? I don’t want to plaster the whole thing in here?

**Comment by tzapu** [comment]
yeah, it doesn t seem to work... dropbox link or similar?

**Comment by scargill** [comment]
Wil this do?

https://www.dropbox.com/sh/kqp8dz4za9zj2kl/AACwlLA0hYGy0i_cDFt32Fjya?dl=0

**Comment by tzapu** [comment]
yup, perfect. 
looked through and i assume this is the non verbose version of the log. could not see any disconnect. 
errors should appear anyway, but could not see those either.
i have 2 thoughts on this:
- maybe it needs some form of keepalive, activity when the phone app is not connected to the server to request data. maybe add an inject/timer to send something to blynk server every 1 minute, indiferent if it is seen or not.
- or maybe it disconnects for some reason and it does not go into retry mode. I have updated to v0.0.18 that ignorese the verbose flag and would print out disconnect messages at all times so we can see if this happens. please update

**Comment by scargill** [comment]
Hi

Ok I’ve updated both installations to the new versions. I already do one second updates sending a flashing led message to the APP regardless of whether it is online or not.

Will let you know tomorrow how things are going – many thanks for your efforts.

Pete.

**Comment by scargill** [comment]
Results of some tests… got up this morning 6.30am – both Spanish and UK installations -  Blynk nodes dead with red “disconnected”. Late last night neither installation had any issues. Neither Node-Red installation has reset since then (as I would have an email from them).

In both cases, restarting node-red restored blynk node.

Here is a folder within it UK and SPAIN folders with node-red logs and BLYNK logs – ignore any reboots before midnight last night – I was updating other software unrelatedto this.

Hope this helps.

https://www.dropbox.com/sh/zue4tjjujjivdt2/AABKb_ftdS5i4_Wxe9hhN17za?dl=0

Pete.

**Comment by tzapu** [comment]
hi, found the disconnection
11 Dec 03:26:52 - [info] [blynk-server:9efc257a.6103d8]  disconnected
11 Dec 00:57:23 - [info] [blynk-server:70d2fa56.8f2d04]  disconnected

is there any chance you could start node-red in verbose mode so we get more data in the log? we re one step closer to this now. i know it s annoying to have to debug... mostly because of the time it needs to happen
mine is still alive, connected and kicking data out btw

**Comment by scargill** [comment]
As I’m using PM2 to autoboot Node-Red  I just have to figure out how to update that setup to add –v   ……. Hmm…

**Comment by scargill** [comment]
Just a reminder that it DID come back up on Node-Red restart (not reboot of Pi)…

**Comment by tzapu** [comment]
yeah, i ve i set it up, when node red goes down, it deletes all old connections and creates new ones, so it would have to come back

**Comment by tzapu** [comment]
look here
http://pm2.keymetrics.io/docs/usage/pm2-doc-single-page/#42-ways-of-starting-processes http://pm2.keymetrics.io/docs/usage/pm2-doc-single-page/#42-ways-of-starting-processes

**Comment by scargill** [comment]
Here’s an interesting thing.

In my Spanish setup, I have TWO Blynk connections – same local server.. ONE of them is disconnected (lasted maybe 24 hours) – but the other is still connected – same server… that indicates to me that the underlying library isn’t at fault here??

Here’s the thing – I went over to the right side of Node-Red where you see the connections – and double clicked on the dead connection to take a look.  At that point you have options to delete, cancel or update. I hit update – and despite not making any changes, the DEPLOY light came on at the top of the screen – I hit that and lo – the Blynk connection came back to life. Has to be a clue in there somewhere.

**Comment by tzapu** [comment]
hi, hitting deploy does in essence what node red restart with do, clears and recreates the connections.
my blynk has now been running and pushing data for 5 days without hickup. 
i need to get a scenario that s more similar to yours.
i am using a single token currently, how many are you using?
do you have the same pin numbers used on different tokens? like v0 used on both project key 1 and 2 and 3 ?
and not sure what else to ask, maybe the issue is in the multiple running dashboards bit...

**Comment by scargill** [comment]
Hi

In England I am using one token with 16 connections…

In Spain I am using 2 tokens, the first has 4 connections – the second has 17 connections.

This morning both of the Spanish connection were broken.

The first one restarted when I hit UPDATE  - DEPLOY

The second one refused to start up even then. However a Node-Red restart brought it back to life.

I’ve no choice but to use 2 connections as I’ve run out of space on the first page. I’ve griped to the BLYNK people about this but to no avail. Their RGB controls are far too big.

**Comment by scargill** [comment]
The UK site which has only one connection – has died again. Node-red restart brought it back…

**Comment by tzapu** [comment]
@scargill any verbose logs that i can look at?

mine has been running non stop now for a week...

**Comment by scargill** [comment]
I will turn verbose on – I’m held back a little here as we’re having the broadband replaces and I’m messing with modems!!   It died again this morning (uk installation with just one connection) – opening the connection – UPDATE  once again restarted it.

Pete.

**Comment by tzapu** [comment]
hi, whenever suits you, just on one server should be fine, no need for both
cheers

**Comment by tzapu** [comment]
finally got a disconnect myself, there doesn t seem to be any reconnect attempt at all. 
i will look into it more, but there s not point now to give me your logs, saw what i needed, thanks

**Comment by Toshibass** [comment]
Would be interested to see what your log looks like (just to compare) my setup has also been semi stable over the last few days, although I have had a few ... 

17 Dec 18:04:52 - [red] Uncaught Exception:
17 Dec 18:04:52 - Error: write EPIPE
    at errnoException (net.js:905:11)
    at Object.afterWrite (net.js:721:19)

mainly when i have been deploying node-red for other reasons nothing to do with n-r-c-b

**Comment by scargill** [comment]
I’m so pleased – not because your excellent program has gone wrong :) but because you now know I’m not going mad.  No, when disconnected I have NEVER seen it try to reconnect… hence the restart of Node-Red. Clearly it’s not the BLYNK server or restarting Node-Red would not make any difference.

Good luck – let me know If you want anything testing.

Pete.

**Comment by tzapu** [comment]
added some more fixes to the blynk library, hopefully the blynk team will merge them quickly
https://github.com/vshymanskyy/blynk-library-js/pull/13

**Comment by scargill** [comment]
Does this mean you have updates or we have to wait for BLYNK to do something?

**Comment by tzapu** [comment]
hi, hopefully i have found what happened and fixed it. but the fix is not in my library is in BLYNK's
so they need to accept my fix, merge it, release new package and then we will all rejoice
i have pushed the changes from my node red lib, but they are rather useless without the above.

**Comment by tzapu** [comment]
@scargill i am also glad i was able to replicate something, hopefully it s the same problem you guys have encountered
@Toshibass sorry, i missed your post, mainly my log would just show a disconnect and no attempt what so ever of reconnect. i did not have any exception or error thrown.

```
2015-12-17T19:41:43.821+0200[App/0]OUT17 Dec 17:41:43 - [info] [blynk-lcd-print:b533ba5a.4acc48] disconnected
2015-12-17T19:41:43.821+0200[App/0]OUT17 Dec 17:41:43 - [warn] [blynk-server:c24f6f52.3db09] key: 3e937439596f4ec68851d17b1b25 disconnected
2015-12-17T19:41:43.823+0200[App/0]OUT17 Dec 17:41:43 - [info] [blynk-write:6cdd0d0.f9322f4] disconnected
2015-12-17T19:41:43.824+0200[App/0]OUT17 Dec 17:41:43 - [info] [blynk-write:be2b6429.41d498] disconnected
2015-12-17T19:41:43.824+0200[App/0]OUT17 Dec 17:41:43 - [info] [blynk-read-event:4f0873ad.b0f78c] disconnected
2015-12-17T19:41:43.824+0200[App/0]OUT17 Dec 17:41:43 - [info] [blynk-read-event:aac13828.553ec8] disconnected
2015-12-17T19:41:43.824+0200[App/0]OUT17 Dec 17:41:43 - [info] [blynk-write:de69917.f21967] disconnected
2015-12-17T19:41:43.824+0200[App/0]OUT17 Dec 17:41:43 - [info] [blynk-read-event:d35491e9.2cab7] disconnected
2015-12-17T19:41:43.824+0200[App/0]OUT17 Dec 17:41:43 - [info] [blynk-read-event:7847c61a.87b838] disconnected
2015-12-17T19:41:43.824+0200[App/0]OUT17 Dec 17:41:43 - [info] [blynk-read-event:c8fca961.370358] disconnected
2015-12-17T19:41:43.824+0200[App/0]OUT17 Dec 17:41:43 - [info] [blynk-read-event:e2704bf4.1d8fb8] disconnected
2015-12-17T19:41:43.824+0200[App/0]OUT17 Dec 17:41:43 - [info] [blynk-write:d0d5770c.2f2a88] disconnected
2015-12-17T19:41:43.824+0200[App/0]OUT17 Dec 17:41:43 - [info] [blynk-write:8bca4506.7435b8] disconnected
2015-12-17T19:41:43.824+0200[App/0]OUT17 Dec 17:41:43 - [info] [blynk-write:f6e91480.0916e8] disconnected
2015-12-17T19:41:43.824+0200[App/0]OUT17 Dec 17:41:43 - [info] [blynk-read-event:596546c1.a69ab8] disconnected
2015-12-17T19:41:43.824+0200[App/0]OUT17 Dec 17:41:43 - [info] [blynk-read-event:831b0c14.7ce4f] disconnected
2015-12-17T19:41:43.824+0200[App/0]OUT17 Dec 17:41:43 - [info] [blynk-write:d9d86191.2627a] disconnected
2015-12-17T19:41:43.824+0200[App/0]OUT17 Dec 17:41:43 - [info] [blynk-write:d05ad2df.2fa53] disconnected
2015-12-17T19:41:43.824+0200[App/0]OUT17 Dec 17:41:43 - [info] [blynk-read-event:a2374d16.5dc8b] disconnected
2015-12-17T19:41:43.824+0200[App/0]OUT17 Dec 17:41:43 - [info] [blynk-write:91200a5e.6edff8] disconnected
2015-12-17T19:41:43.824+0200[App/0]OUT17 Dec 17:41:43 - [info] [blynk-write:b5f74e59.4a08b] disconnected
```

**Comment by scargill** [comment]
I think you’d better have a chat with the people at Blynk – I asked them to incorporate your changes and this is the result..

  http://community.blynk.cc/user_avatar/community.blynk.cc/dmitriy/45/226_1.png 

 http://mandrillapp.com/track/click/30526449/community.blynk.cc?p=eyJzIjoiUXFkNGJnMjFlbDR4MVhqUFR6VjJCS2tobEc4IiwidiI6MSwicCI6IntcInVcIjozMDUyNjQ0OSxcInZcIjoxLFwidXJsXCI6XCJodHRwOlxcXC9cXFwvY29tbXVuaXR5LmJseW5rLmNjXFxcL3VzZXJzXFxcL2RtaXRyaXlcIixcImlkXCI6XCIzNjgzNGVmYjg0ZGY0YzUwYmVlOWEwYjA3YmEyYmY2MFwiLFwidXJsX2lkc1wiOltcIjg4ODliZmYyM2Y0OTI2MjJlMDY2MjcwNDcwZjFhOGU0M2I0NTQ0NmRcIl19In0 Dmitriy Co-Founder 
December 18 

Could you please explain what fix do you mean?

To respond,  http://mandrillapp.com/track/click/30526449/community.blynk.cc?p=eyJzIjoiU25NelVURGlSTVZad2ZaaE12UkxERHczbmVBIiwidiI6MSwicCI6IntcInVcIjozMDUyNjQ0OSxcInZcIjoxLFwidXJsXCI6XCJodHRwOlxcXC9cXFwvY29tbXVuaXR5LmJseW5rLmNjXFxcL3RcXFwvYmx5bmstZm9yLW5vZGUtcmVkXFxcLzMyMDRcXFwvMlwiLFwiaWRcIjpcIjM2ODM0ZWZiODRkZjRjNTBiZWU5YTBiMDdiYTJiZjYwXCIsXCJ1cmxfaWRzXCI6W1wiOGIzZTEzYTE1ZWRjYjljMDU1MTM3YWI5ODBiOTJlNGM0MGUwYTAxZlwiXX0ifQ visit the topic in your browser.

**Comment by scargill** [comment]
Well, this morning the node-red blynk icons were all stopped again…

Here’s the link to my various logs for what it is worth..

https://www.dropbox.com/sh/k39z2bnsdll4azw/AAAPSwIBuHH6GXclwn-yC64La?dl=0

I restarted the Blynk server – no difference (I don’t think it had stopped).

I looked at the config node – hit update on that – nothing….

I restarted node-red – all is well….  Until the next time it stops – nothing else is stopping – just the blynk stuff..

Pete.

**Comment by tzapu** [comment]
aww, crap..
will look at it

**Comment by tzapu** [comment]
any node red logs as well? 
out of curiosity, when the icons were off, was the mobile app not updating as well?

**Comment by scargill** [comment]
The PM2 log file is node-red (strange indeed)… you know I didn’t check the APP – I’ll have to do that next time – ok, until then….

**Comment by tzapu** [comment]
nothing in pm2.log from node-red directly
i think there s another service-name.out.log or similar that pm2 makes that contains the output from the service

**Comment by tzapu** [comment]
maybe it s also worth listing what packages you have installed, what versions for blynk
something like 
`npm list -g --depth=1|grep blynk`

**Comment by scargill** [comment]
Erm.. nothing happened when I did that – just came back to the prompt ??

**Comment by tzapu** [comment]
weird, maybe it s different on a rpi

i get this

```
npm list -g --depth=1|grep blynk
├─┬ node-red-contrib-blynk@0.0.19
│ ├── blynk-library@0.0.25

```

did you install node-red globally '-g' or locally to a folder?

**Comment by scargill** [comment]
Locally.

I install my stuff generally by sitting in the .node-red directory and doing npm install xxxxx

I just tried that command with and without the –g and I get nothing… strange…-library

If I do

Npm view blynk-library version

I get 0.0.25 so at least that is up to date

And for npm view node-red-contrib-blynk version

I get 0.0.19

Pete.

**Comment by scargill** [comment]
A thought – my friend David who has also seen the blynk node just stop working as happens for me – seems to think that one reason for that might be too fast MQTT inputs  - can you think of a reason why that might be? Could it be the cause of the issue…

Here is David’s input..

Just a thought....

Have you got MQTT input nodes in Node Red that are flooding your Blynk nodes?

I just got Blynk disconnect problems using the bundled (i.e. older) version of Node Red on my Pi Zero. That's the first time I've had it on that version.

It only started after I had added a DHT22 sensor to my ESP board in the garage and also set up a new ESP board in the living room with a DHT22 sensor.

A bit of backtracking revealed that it was ok if I had the garage one running but as soon as I had this:

for the living room one, it caused the Blynk nodes to disconnect. If I removed those MQTT nodes, or even kept them but changed the topic to one that was not being published to, then it was ok.

The garage one (which was NOT causing disconnects) is doing the publishing of sensor data and also is subscribed to some incoming MQTT topics.

But the living room one (that was causing the crashes) was ONLY sensing the DHT22 data and publishing it. Nothing else. So it was sending several MQTT packets per second. So I stuck a 

delay (1000);

in the void loop section and as far as I can tell so far that seems to have fixed it.

Could your problems be caused by something similar??

**Comment by tzapu** [comment]
there was one other user that mentioned mqtt issues bringing down blynk as well, but i could not replicate that either 
i believe david is using blynk.cc would would restrict the number of requests per second so in his case, a disconnect and/or delay would be sure to happen if he were to flood them.

but this also does not apply in case of your local server, where the requests should be set much higher.
my prev version test is still going strong, had a server fail and it kept trying to connect, i was able to see in the log the reconnection attempts. In the end it wasn t successful because i had to mess with dns so it lost the plot... but it did try to reconnect for a few hours quite feverishly

**Comment by autolycus2000** [comment]
Yes, that's right - I'm using blynk.cc.

What I have established so far is as follows:
1. I had Node Red and Blynk running for several days with no problems. I had several flows running which were SENDING out MQTT broadcasts in response to button presses on Blynk.
2. As soon as I added some MQTT that were INPUTTING into Node Red and then sending their payload to value displays on Blynk I started to get random disconnects. See attached JPG for an example of the kind of flow I am talking about.

![flow](https://cloud.githubusercontent.com/assets/16492470/12055540/15b3d770-af26-11e5-9985-428c321c3328.JPG)
1. I have multiple screens/projects in Blynk. Some of them are receiving these incoming MQTT messages and some are not. When the disconnects happen it is ONLY the projects that receive incoming MQTT messages which disconnect. The other projects all stay running fine.
2. Once the disconnect happens, my Node Red log file fills with messages like this:

30 Dec 10:45:04 - [info] [blynk-server:7cffee53.83001] key: xxxxxxx disconnected
Connecting to SSL: cloud.blynk.cc 8441
Connected
Disconnect blynk

(where xxxxxx is the key of the project that the MQTT data is being sent to)

and then after several pages of that it crashes completely with:

30 Dec 10:46:24 - [red] Uncaught Exception:
30 Dec 10:46:24 - Error: write EPIPE
    at errnoException (net.js:904:11)
    at Object.afterWrite (net.js:720:19)

I have tried reducing the frequency with which the data is sent from the incoming MQTT message to the Blynk node. Even reducing it to every 15 seconds it still causes disconnects.

Does that help at all?

David.

**Comment by scargill** [comment]
As you know I don’t think I’m getting those messages but I’m still getting occational disconnects. I hope the info David is providing helps here… would be great to make this bulletproof…

Pete.

**Comment by autolycus2000** [comment]
Especially as the guys at Blynk plan to start charging us for it in the next couple of months.

**Comment by scargill** [comment]
Not until it works – and until they announce what they plan to do about new controls – the RGB control is still WAY too big to be of any use…  I don’t know what they think we’re going to do with this but I have 4 RGB lights in one room!

:)

**Comment by tzapu** [comment]
guys, thanks for this.
i will be back on monday in full working order and will try to build more test setups with mqtt as well to see if i can replicate it that way.
it s so so annoying that i can t properly replicate this...my server is still going strong

**Comment by autolycus2000** [comment]
Thank you!

I've just had another disconnect and here are the details in case it helps you replicate or fix the problem.

I have got 10 flows in Node Red. The three that are significant for this example are called: Garage, Living Room, and Sensors.

I have 5 tokens/projects running in Blynk. The ones that are significant for this example are called Garage, Living Room, and Sensors.

On the Garage Blynk project there are a switch and an LED that are used to turn a light on/off and display it's status. Also a digital readout for temperature and one for humidity. And two history graphs - one for temperature and one for humidity.

On the Living Room Blynk project there are various buttons and sliders for controlling lights, as well as two digital readouts - one for temperature and one for humidity.

And on the Sensors Blynk project there are 4 digital displays which display garage temperature, garage humidity, living room temperature, and living room humidity. Plus two history graphs - one for the two temperature sensors and one for the two humidity sensors.

In Node Red the Living Room flow talks to the Living Room project in Blynk - but ONLY to the buttons and LEDs It does not interact at all with the digital displays or graphs.

The Garage flow talks only to the Garage project and only to the buttons and LEDs.

The Node Red flow called Sensors takes readings from DHT22 sensors via MQTT nodes and sends these on to EmonCMS, Grovestreams, Dweet, and the Blynk nodes for the various digital displays and graphs referred to above. So that means the Sensors flow is talking to all three of the Blynk projects I referred to above.

The disconnect that just happened ONLY affected the Garage project in Blynk. So, my Living Room flow - screeshot below - is all still working fine with all Blynk nodes connected.

![image](https://cloud.githubusercontent.com/assets/16492470/12074879/9b6b9958-b164-11e5-8c02-a7c2b8291dcc.png)

On my Garage flow the Blynk nodes (which are all either LEDs or buttons) have all disconnected:

![image](https://cloud.githubusercontent.com/assets/16492470/12074885/dd10be56-b164-11e5-83f1-ea403001f320.png)

And on my Sensors flow, the nodes that are part of the Garage project are disconnected but all the others are ok:

![image](https://cloud.githubusercontent.com/assets/16492470/12074887/094ce418-b165-11e5-88b3-5e4d6fc767c9.png)

I am convinced that the problem is related to sending data from Node Red to the digital display widgets and/or the history graph widgets. I say this because:
a) I didn't get disconnects before I started using these widgets
b) the disconnects only ever affect projects that include these widgets. My other Blynk projects (e.g. the Living Room one mentioned above) do not use digital displays or graphs and do not suffer from disconnects.

Hope this helps. Feel free to ask any questions.

David

**Comment by autolycus2000** [comment]
UPDATE: In the past week I have had 2 or 3 disconnects on flows/projects that do NOT use digital display or history graph widgets. So my theory in the post above has now been shown to be incorrect.

**Comment by tzapu** [comment]
guys, sorry about taking so long, i have activated authentication on my node red, and have lost access to it, no matter what i tried with bcrypt and hashes and all the documentation, it doesn t let me in. 
it pissed me off so much that i didn't even feel like looking at it for the past few days.
anyway, will try some more, if not reinstall.

**Comment by tzapu** [comment]
with the latest updates from the Blynk team, i built a (very alpha) contrib node that uses websockets instead of the blynk tcp transport.
caveats:
- first build, very alpha
- only does the basics now, read event, write event, write
- you need a local blynk server installation, with the latest binary for it to work
- no ssl yet and no cloud.blynk.cc until they allow websockets in production

Now for the good stuff:
- this should allow me to implement stuff directly and it also has a smaller chain of dependencies, so, i may be able to intervene/fix stuff better going down this route.
- It is a replacement for the other contrib, but they can coexist, like on my testbed (3 projects, one logacy, 2 websockets, sending continuously)

<img width="1061" alt="screenshot 2016-01-21 09 08 50" src="https://cloud.githubusercontent.com/assets/2983312/12473856/a8cd1554-c01e-11e5-8f12-f70e2a3b351a.png">
- in my testing at least, it s really really really fast
- seems to reconnect, but time will tell, i did not have any problems with the other build either

is anyone willing to give it a go? to install you need to do
npm install https://github.com/tzapu/node-red-contrib-blynk-websockets
in your node-red folder
cheers

**Comment by Toshibass** [comment]
Hi Tzapu sounds very interesting, problem is I am using cloud not local server, I suppose I could try installing local server and then give it a go but it won't be for a couple of weeks as I am on vacation , I am currently still getting disconnect issues  which is very irritating.

**Comment by scargill** [comment]
Keep going for me as I prefer local server.

**Comment by tzapu** [comment]
well, i have it going for 36+ hours now, 4 projects connected (1 with old contrib, 3 with webscokets contrib) passsing a lot of data when app is open, and still passing data every 30s on around 10vpins when the app is closed...
so far so good

i will introduce mqtt to it as well, to see what happens 

@Toshibass from what the blynk guys told me, websockets will get included in the cloud version as well, they just need to optimize it and check it some more to make sure it s stable and not bringing their servers to the knees

@scargill if you could give the node a try it would be very good, you could even just replace a few of your current blynk nodes with the new ones, on the same project, and see if some disconnect and others not

**Comment by scargill** [comment]
MQTT – you have my interest!!

**Comment by tzapu** [comment]
@scargill sorry for that, i meant add some mqtt nodes to my test setup because @autolycus2000 i think mentioned mqtt nodes in his flows might have had something to do with disconnects

**Comment by scargill** [comment]
Oh ok. Well as soon as I get back in my office next weekend I’ll give this another attempt….   Thanks for putting in the effort on this – much appreciated.

Pete.

**Comment by Toshibass** [comment]
Sounds promising, it may be of interest, I have 2 projects 1st has 37 blynk nodes and 17 mqtt nodes, 2nd project has 30 blynk nodes and 32 mqtt nodes, originally I had both projects on 1 instance of node-red on a raspberry pi2 however because of the numerous disconnect issues I split the 2 projects onto 2 raspberry pi’s (I still use just 1 mqtt broker) the result was a noticeable improvement on the number of disconnect issues, although I still do get some disconnects were I need to at least re-deploy or worst case restart node-red.

One other point my blynk nodes are not passing data continuously, I have built the flow so when the blynk app opens it triggers a “blynk write even node” which is connected to mqtt which starts the sending of data to “blynk write nodes”, some send values to “value displays” most send data write the state of gpio’s to virtual buttons on the app, the same virtual buttons when pressed send data to “write event nodes” to toggle the value of the relevant gpio, when the app is closed then no data is passed between node-red the blynk.

In other words its pointless trying to send and receive data between rode-red and blynk if the app is not open.  

Toshi

**Comment by tzapu** [comment]
yup, i am sending data continuously more as load testing and to get nice full history graphs really
i just want to see if connections hold

**Comment by tzapu** [comment]
well, good news and bad news
bad news is that it seems the original version of the blynk contrib does fail faster when mqtt nodes are involved. Honestly, i don t know how that can happen as all nodes are sandboxed as far as i know.
it also tries to reconnect, continuously but seems to drop the connection immediatelly. 

```
Connecting to TCP: server:port
Connected
Disconnect blynk
25 Jan 00:12:30 - [info] [blynk-server:de574db2.efd6d] key: 3e937439596key25bde9 disconnected
```

it reconnected after a redeploy.

the good news, at least i hope so, are that the websockets version of this stayed connected and working even though the other one was down... i m having more and more of an incline to scrap the propriety protocol one and just continue work based on websockets...

this is how it looked, lighter green = websocket nodes
<img width="1016" alt="screenshot 2016-01-25 08 08 17" src="https://cloud.githubusercontent.com/assets/2983312/12547920/b4efa956-c35b-11e5-8790-050efc6bdeb3.png">

**Comment by scargill** [comment]
Tzapu - I'm so pleased you've been able to (a) replicate the issue and (b) find a way around it because frankly I could not devote more time to it when it was continually failing on me and I use a LOT of MQTT.. The problem you have is exactly what I had - I had to restart Node-Red - but I never had to restart the computer itself - OR the Blynk server.

I'll take a look at this when I'm back in the UK next week - I think you might be right about scrapping the original as others obviously had similar issues - I know David Miles did (I think he is on holiday I think which is why he is quiet).

When you have this ready to go with a little documentation so I don't have to harrass you - let me know and I'll give it a go.

Great stuff - and thanks for progressing this. 

Pete.

**Comment by scargill** [comment]
So – is a working websockets version up and available???

**Comment by scargill** [comment]
Ok.. .what am I doing wrong.

I put the latest server on my PI – and installed… and got the app talking to it – new account – logged in ok.

I put 2 nodes on the Pi Node-Red – the new WS types…

And I copied the KEY into the connection and 

The APP says my project is not connected and Node-Red says..

node-red-contrib-blynk-websockets:websocket.errors.missing-conf

I have no config file yet, just the admin, client and server JS files – latest JAVA etc..

What am I missing?

**Comment by scargill** [comment]
Someone put me out of my misery – I have the server working – the app working – cannot get the node to talk to the server.. what should I have in there…

**Comment by scargill** [comment]
Right –mine is working – but wrong – in your DEFAULT you have blah…./ws   but to get it to work Toshi showed me that it needs /websocket   - might want to fix that.

Only had it running for an hour but had 2 brownouts which reset the router – and it has survived that – the wifi and mains are ATTROCIOUS here in my temporary home near Barcelona (project) but it has survived so far – let’s see how well this WS  works…

TZAPU – are you giving it a good hard test with constant MQTT requests?

Pete.

**Comment by scargill** [comment]
Ok, question – got the new setup running…

If I make a slider and arrange 2 injects – 0 and 200 – I can send the slider going back and forth.. great.

But if I set the slider to 0 with the button… and STOP the app – and press the 200 button and THEN START the app – it does not seem to catch up – as I understood it the info is held in the server regardless of whether the APP is running or not.

So why isn’t the app catching up when you start it again???

Bug? Misunderstanding?  Really that needs to work – you have no way to tell the APP has just started and needs refreshing so it needs to pick up the values from the server…

Pete.

**Comment by tzapu** [comment]
hi, 
glad you got it working,  i was fast asleep :)

sorry for the lack of documentation and everything, but i didn t want to put more time in it before knowing it s a valid path to go down on (which it seems it is, yeey)
yes, every timer i have on my test pages, one half a second, and a few 30s ones, i m actually proxying through mqtt to the flow for testing, so every few blynk requests theres an mqtt request

regarding your slider thing, as far as i understand, the blynk guys added a sync ability recently, but looking at their documentation it seems to do the oposite of what you need:

```
State syncing
Blynk allows you easily restore hardware state from widget values in case your hardware was reset or disconnected. We call this feature state syncing. Here is simple code snippet :
BLYNK_CONNECTED() {
  if (isFirstConnect) {
    Blynk.syncAll();
  }
}
The syncAll command set all virtual pin values based on what is currently set on the app. In other words every virtual widget will generate BLYNK_WRITE event on hardware side.
```

while testing, i for one did not see any sync messages go to node -red

the guys seem flexible enough, and seem to be fulfilling any request that makes sense, so maybe we just need to ask them for a Blynk App Open event, on which everyone could just write the latest state to the interface.

if this thing is more reliable for you as well, i ll start going through all the motions, publish to npm, touch up, add docs, implement the rest of the widgets, and so on

cheers

**Comment by scargill** [comment]
I got up this morning and… all read – disconnected.. but then I noted that the server was down – I suspect a reboot and I’ve not yet set it in the startup.

I started the server and the nodes came back on… no problem.

So it would seem the only issue is…

You have a button – you turn it on and off via Node-Red  - fine…  Turn it ON. STOP the App… turn it OFF… Start the App. It’s state is not updated.

Is this the fault of the APP, the SERVER or the Node-Red Interface..

Perhaps you can talk to the Blynk people?

I’ll ask also.

Pete.

**Comment by tzapu** [comment]
that s good news then
i ll talk to them as well, there s a slack channel now, so this should be easier

**Comment by tzapu** [comment]
@scargill spoke about your issue to the node red guys, and they said it should work.
tested it myself like this
https://www.dropbox.com/s/tjd69glbs2hse2o/Screenshot%202016-01-27%2012.15.19.png?dl=0

V30 is a small slider in my app

and i can report it does work, the slider changes even when app in background or completly killed

i tested on ios...

btw, this only works for virtual pins

**Comment by scargill** [comment]
The blynk guys came back with this..

Using the WRITE node – and feeding 200 or 0 to the slider, I see it move right and left… but if I STOP the app and change the state with node-red – the change does NOT reflect in the APP when I start it.

I am using a Nexus 10 Android tablet – Android 5.1.1  and a phone with an older version 4.2 Android – the result is the same

I downloaded new server 0.12.4 .jar,  admin 0.12.0.jar and the latest version of the App yesterday.

With the above tablets and using values 0 and 200 for the slider, they work perfectly until I stop the app – make a change then start the App – the App DOES NOT update…

Pete.

 http://mandrillapp.com/track/click/30526449/community.blynk.cc?p=eyJzIjoiN0xDTnluRjlvMmx5VUZWanAxNTNOMWEwNmRRIiwidiI6MSwicCI6IntcInVcIjozMDUyNjQ0OSxcInZcIjoxLFwidXJsXCI6XCJodHRwOlxcXC9cXFwvY29tbXVuaXR5LmJseW5rLmNjXFxcL3VzZXJzXFxcL3NjYXJnaWxsXCIsXCJpZFwiOlwiMjE5ZDViZTUwODVhNGE3MDhmOTg4ZDllNDMxMDViNjJcIixcInVybF9pZHNcIjpbXCI4NjNlYmEwMDM3NGEzYWM2YWY1ODk0NmUxMDg0NzI0YTM2NTlkYTc3XCJdfSJ9 @scargill

Hello. You need to use Blynk.virtualWrite on button press if you need sate syncing.
For instance, let's assume you have button on V1 pin. And on your hardware you have interrupt function for physical button. So when you click physical button you need also call

Blynk.virtualWrite(V1, valueToSetOnApp);

This works for sliders also.

To respond,  http://mandrillapp.com/track/click/30526449/community.blynk.cc?p=eyJzIjoianlqZFllQVZ4cl9pZTNOaTgtNThuVm1ORl9NIiwidiI6MSwicCI6IntcInVcIjozMDUyNjQ0OSxcInZcIjoxLFwidXJsXCI6XCJodHRwOlxcXC9cXFwvY29tbXVuaXR5LmJseW5rLmNjXFxcL3RcXFwvd2ViLXNvY2tldHMtdXBkYXRlLW9uLXN0YXJ0aW5nLXRoZS1hcHBcXFwvMzkxNFxcXC8yXCIsXCJpZFwiOlwiMjE5ZDViZTUwODVhNGE3MDhmOTg4ZDllNDMxMDViNjJcIixcInVybF9pZHNcIjpbXCI4YjNlMTNhMTVlZGNiOWMwNTUxMzdhYjk4MGI5MmU0YzQwZTBhMDFmXCJdfSJ9 visit the topic in your

**Comment by scargill** [comment]
This is what I just sent off to the BLYNK people…  I can replicate this reliably.

I should clarify…

If I KILL Blynk – ie stop the application running  - make a change in Node-Red – and restart Blynk – then all is well – any changes made while Blynk was not running are reflected in the controls.

If on the other hand I just STOP Blynk (with the control top right of screen) – then make a change – and then start, NO changes are reflected in the APP.  My guess then is that this HAS to be an issue with the APP because obviously the server is storing the data.

So – if I use the rectangle top right of the App to stop it running… make a change elsewhere, ie. Fire a value at the slider from Node-Red – then START the app again using the right-arrow top right of screen – then no updates are shown.

Can you replicate this. I’m using Android.

Pete.

**Comment by tzapu** [comment]
cool,

maybe it s an android thing,
works perfectly for me on ios it seems

alex

**Comment by scargill** [comment]
Would it be possible to have a NAME for the connection – and would it also be possible for them to be global – I noted when switching tabs that I had to create a new connection for each tab… whereas for example an MQTT connection is available across the range.

The reason for naming – is that the source of the connection may be in more than one place.

**Comment by tzapu** [comment]
i ll add name 
for global scope, there s been a change in node-red, you ve now got a scope dropdown in the top right corner of any config window
https://www.dropbox.com/s/kn43ktwwkqovm9c/Screenshot%202016-01-28%2019.06.38.png?dl=0

**Comment by scargill** [comment]
HEY – isn’t that neat……

Ok, will look forward to a name … much better – and thanks for the education!

Pete

**Comment by tzapu** [comment]
now on npm 
https://www.npmjs.com/package/node-red-contrib-blynk-websockets
with name for server label
will slowly move all of the current one to the websockets one and start updading docs and so on.

**Comment by scargill** [comment]
Hi

Everything continues to work well – you’ve done a good job..

Question – it might just be me but..

On powerup of Node-Red if I wanted to see what state a button is currently in – i.e. 1 or 0 -  I can’t see how I would do that – is there any way to interrogate a button?

**Comment by tzapu** [comment]
it looks like there might be a way to do it here
https://github.com/blynkkk/blynk-library/blob/master/examples/More/SyncPhysicalButton/SyncPhysicalButton.ino

i ll have to look into it

for new issues and problems please use a new issue on the websockets project, this will be cleaned up and made obsolete soon

cheers
